### ISSUE TITLE
train.py keep waiting when using multi gpus to train Transformer

### ISSUE THREAD

**Issue by huhk-sysu** [OpenNMT/OpenNMT-py#954]
I'm training the Transformer model on multi-gpu and I have done `export CUDA_VISIBLE_DEVICES=0,1,2,3`. Below is my command I run(almost the same as the one in [FAQ](http://opennmt.net/OpenNMT-py/FAQ.html#how-do-i-use-the-transformer-model)).

```bash
python train.py -data ./data/wmt2014/processed -save_model ./saved_models \
        -layers 6 -rnn_size 512 -word_vec_size 512 -transformer_ff 2048 -heads 8  \
        -encoder_type transformer -decoder_type transformer -position_encoding \
        -train_steps 200000  -max_generator_batches 2 -dropout 0.1 \
        -batch_size 4096 -batch_type tokens -normalization tokens  -accum_count 4 \
        -optim adam -adam_beta2 0.998 -decay_method noam -warmup_steps 8000 -learning_rate 2 \
        -max_grad_norm 0 -param_init 0  -param_init_glorot \
        -label_smoothing 0.1 -valid_steps 10000 -save_checkpoint_steps 10000 -gpuid 1 2
```

When I only use one GPU (e.g. set -gpuid 1), everything is ok, the training goes well and there're many logs. However when I use multi gpus to train (e.g. set -gpuid 1 2), the process seems stucked. It just keep waiting and nothing happens. I waited for several minutes but found nothing printed to the console. I used `top` command and found there were 4 python process running. I stop the process using `CTRL+C` and found the following traceback:

```
Traceback (most recent call last):
  File "train.py", line 40, in <module>
    main(opt)
  File "train.py", line 25, in main
    multi_main(opt)
  File "/home/huhk/models/omnt/onmt/train_multi.py", line 38, in main
    p.join()
  File "/home/huhk/miniconda3/lib/python3.6/multiprocessing/process.py", line 124, in join
    res = self._popen.wait(timeout)
  File "/home/huhk/miniconda3/lib/python3.6/multiprocessing/popen_fork.py", line 50, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/home/huhk/miniconda3/lib/python3.6/multiprocessing/popen_fork.py", line 28, in poll
    pid, sts = os.waitpid(self.pid, flag)
```  

I'm new to here so I'm not sure if there is anything wrong with my actions. Could anyone tell me what the problem may be? BTW, I'm using pytorch 0.4.1 and the lastest OpenNMT(I just downloaded it yesterday). I'm waiting to offer more information if needed.

Thanks for your help!

**Comment by vince62s** [comment]
look at the end of FAQ
Do you support multi-gpu?
Yes ! First you need to make sure you export CUDA_VISIBLE_DEVICES=0,1,2,3 Then use -gpuid 0 1 2 3 If you want to use GPU id 1 and 3 of your OS, you will need to export CUDA_VISIBLE_DEVICES=1,3 then use -gpuid 0 1

**Comment by huhk-sysu** [comment]
Thanks for your reply @vince62s . I did notice this, and I have tried `export CUDA_VISIBLE_DEVICES=0,1,2,3` as I mentioned above but it failed.

For the case I'm using two gpus, I also tried `export CUDA_VISIBLE_DEVICES=1,3` and use `-gpuid 0 1` but there isn't anything happen yet.

![image](https://user-images.githubusercontent.com/14769033/45419262-5b31de80-b6b9-11e8-8cdb-c351efbacbbe.png)
![image](https://user-images.githubusercontent.com/14769033/45419281-6edd4500-b6b9-11e8-8063-31f3b7aa075f.png)

**Comment by vince62s** [comment]
use -report_every 10
can you confirm whether your GPUs are actually working or not ? "watch nvidia-smi" in a CLI

**Comment by huhk-sysu** [comment]
OK. I have checked my CUDA setting again.
![qq 20180913010223](https://user-images.githubusercontent.com/14769033/45441208-145fdb00-b6f1-11e8-9ba7-2700501feff4.png)


When I use "-gpuid 0 1" to train, the nvidia-smi report "No running process found", which may suggest the training doesn't start.

When I use single GPU, no matter what its index is, e.g."-gpuid 0" , "-gpuid 1", "-gpuid 2", the nvidia-smi always report I'm using the 1st GPU (index starts from 0). 

![snipaste_2018-09-13_01-03-55](https://user-images.githubusercontent.com/14769033/45443576-b682c180-b6f7-11e8-9f79-41509ce0c656.png)

**Comment by vince62s** [comment]
what version of cuda and nccl do you have ?

**Comment by huhk-sysu** [comment]
I'm using CUDA 8.0 and I don't have an NCCL installed. I used to train models on a single GPU so I have never used it. 
Maybe I made a big mistake. Is NCCL necessary for training models on multi gpu?  If so, I have to apologize to you for my disturbing.

**Comment by vince62s** [comment]
yes it is. I'll make it clearer on the home page.

**Comment by huhk-sysu** [comment]
Now I haved install NCCL 2.2.13-1+cuda9.0, but the problem is still exist.

**However finally I solve the problem by replacing line 24 at omnt/utils/distributed.py:**

```python 
dist_init_method = 'tcp://localhost:{port}'.format(port=random.randint(10000, 20000))
```

with
```python
dist_init_method = 'tcp://localhost:12345'
```

![image](https://user-images.githubusercontent.com/14769033/45474113-6fceaf00-b76b-11e8-9068-aae8d353d578.png)

I guess the reason is that, in the first way, the `dist_init_method` may be different when `multi_init` being called twice, and the two process can't communicate.

**Comment by vince62s** [comment]
@alphadl did you test your patch ? I merged it thinking it was working.
get back to me or I will revert it, thanks !

**Comment by vince62s** [comment]
okay, looking at it of course it can't work.
the random thing has to happen before starting each process .... @alphadl I am reverting your commit and we will do it differently.

**Comment by vince62s** [comment]
@huhk-sysu I reverted the wrong commit. You can git pull.
please confirm it works for you now.
thanks.

**Comment by huhk-sysu** [comment]
@fotwo This may due to the 2 main training processes tring using the same address(`localhost:10000` currently). 

>  random thing has to happen before starting each process ....

It seems the static port 10000 is temporary and they will make it random later.

**Comment by fotwo** [comment]
OK, I see, thanks.

**Comment by Curry1201** [comment]
Hi @fotwo , I have encountered the same problem as you. How do you solve it specifically?
The problem is as follows：
RuntimeError: Address already in use at /pytorch/torch/lib/THD/process_group/General.cpp:17

**Comment by dimitarsh1** [comment]
Hello all, 

I have a similar issue with both Transformer and LSTM on a 4-GPU with 1080Tis. So, basically it start and prints 
[2019-03-25 16:37:57,076 INFO] Loading dataset from /media/.../data/ready_to_train.train.0.pt, number of examples: 881664

and then it waits forever. 

If I restart the computer, it works normally - it loads, runs through the steps and  so on. If interrupt the process it hangs. Also, sometimes it hangs on loading the validation dataset. I assume some address or memory is not freed if the training process is interrupted. Any idea how to mitigate this?

Thanks, 
Cheers, 
Dimitar

**Comment by vince62s** [comment]
Post a new issue with cmd line
